### ISSUE TITLE
[MENFORCER-185] Require Release Dependencies ignorant about aggregator build

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#689]
**[Thomas Diesler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tdiesler)** opened **[MENFORCER-185](https://issues.apache.org/jira/browse/MENFORCER-185?redirect=false)** and commented

If A depends on B it is ok for A-1.0.0-SNAPSHOT to have snapshot dependency on B-1.0.0-SNAPSHOT if B was build before A during the same reactor build.

Using the requireReleaseDeps rule it seems that SNAPSHOTS are generally not allowed even when they belong to the same project and were built during the same reactor build.

We have a complex project with 100+ modules. I want to enforce that no module has dependencies on project SNAPSHOTS that were not included in the build. In such case A would use a stale version of B that happened to be available in the local/remote maven repository.

---

**Affects:** 1.3.1

**Attachments:**
- [MENFORCER-185.patch](https://issues.apache.org/jira/secure/attachment/12838903/MENFORCER-185.patch) (_116.46 kB_)
- [seuss.zip](https://issues.apache.org/jira/secure/attachment/12716533/seuss.zip) (_1.87 kB_)

**Issue Links:**
- [MENFORCER-186](https://issues.apache.org/jira/browse/MENFORCER-186) Create a rule for the reactor (ReactorModuleConvergence)
 (_**"depends upon"**_)

**Remote Links:**
- [GitHub Pull Request #34
](https://github.com/apache/maven-enforcer/pull/34)

8 votes, 11 watchers

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

See MENFORCER-186 which is exactly what you need. I'm currently working on such rule.

**Comment by jira-importer** [comment]
**[Curtis Rueden](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ctrueden)** commented

+1. My group's projects also use requireReleaseDeps, which works great for single-module projects, but not multi-module ones without hacky exclusions. I attached a nearly MCVE for illustration.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

So first thanks Curtis for your feedback. So after taking a look into that i found some things we need to mention.

The [documentation](http://maven.apache.org/enforcer/enforcer-rules/requireReleaseDeps.html) says that by default the `onlyWhenRelease` configuration element is set to `false` which means the check will be done always. Furthermore the documentation also says the default for `failWhenParentIsSnapshot` configuration element is set to `true` which means it will check the parent always. These are the reasons why the example build fails.

I have created a [github project](https://github.com/khmarbaise/menforcer/tree/master/menforcer-185) with the above example and enhanced it a little with an dependency to `ham` in `green-egs` module but with the wrong version which is not from the current reactor. The newly implemented rule `reactorModuleConvergence` will exactly show this problem.

```
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.ReactorModuleConvergence failed with message:
Reactor modules contains dependencies which do not reference the reactor.
There is a problem in your reactor.
 module: seuss:green-eggs:jar:0.0.0-SNAPSHOT
    dependency: seuss:ham:0.0.1-SNAPSHOT
```

This is of course not the check for SNAPSHOT dependencies what you like to do.
So this looks to me that we need to (re)implement the `requireReleaseDependencies` rule in a complete different way...cause the doc stated:
??This rule checks the dependencies and fails if any snapshots are found.??
Based on that i would interpret that in the way to check only the dependencies and check if they are not part of the reactor to fail.

PS.: The title of the issue is a little bit misleading, cause we are talking about a multi module build and not about an aggregator build.

**Comment by jira-importer** [comment]
**[Curtis Rueden](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ctrueden)** commented

Thanks for the quick reply, Karl. I want to clarify that this is a fringe issue for my group's projects, since we have switched almost completely to single-module projects in separate Git repositories. We have only a couple of multi-module projects left that get bitten. That said, I agree with the OP that if you have a 100-module build, and it is important that it have only release dependencies outside the reactor, it would be awesome if this rule could support that use case. My personal tack would be to add another boolean property `allowSnapshotsInReactor` which defaults to `false` for backwards compatibility, but can be set to `true` to get the desired behavior.

Looking at the source, I wouldn't think a full rewrite is required, but rather some case logic carefully injected in a couple of places. For example, in `RequireReleaseDeps.java`, in the `if ( failWhenParentIsSnapshot )` block, do not throw the exception if `allowSnapshotsInReactor` is set and the parent is part of the multi-module build. And in the `checkDependencies` method, do not add the artifact to the `foundSnapshots` structure if the artifact is part of the multi-module build.

I'm certainly no expert on the codebase though, so perhaps it is more complicated than that.

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

I have attached a working patch for this, see `MENFORCER\-185.patch`. It passes all 9 ITs I added with Maven 3.0.5 and 3.3.9 (on Ubuntu & Windows). However, it requires a signature change of an abstract method (namely `AbstractBanDependencies#checkDependencies`), so it would break current implementations. I don't see a good way around that, so I think this will have to wait for a major upgrade.
